cli: -netinfo quick updates/fixups for 0.21

jonatack opened the pull request:
Quick fixups and updates for v0.21.0:

- [x] handle larger BIP155 `addrv2` addresses
- [x] add Signet chain
- [x] add an additional space between the `net` and `mping` columns; add missing `tinyformat` and `algorithm` headers
- [x] s/uptime/age/ per 0xB10C suggestion, and make the column auto-adjusting variable width
- [x] display `-` for oversized mping/ping times like `1.17348e+06`, as reported by practicalswift

Edit: removed the release note commit, as this PR was not merged before the notes were moved to the wiki. It's here:
```
- A new `bitcoin-cli -netinfo` command returns a network peer connections
  dashboard that displays data from the `getpeerinfo` and `getnetworkinfo` RPCs
  in a human-readable format. An optional integer argument from `0` to `4` may
  be passed to see various levels of detail. (#19643)
```

jonatack committed:
cli -netinfo: various quick updates and fixes

- add new signet chain

- update change "uptime" column name to "age" per suggestion by 0xB10C (Timo)

- add an additional digit to mping field width

- change m_networks_size from size_t to uint8_t, as size_t was a holdover
  from m_networks_size being defined as size_t m_networks.size() in a draft

- order Peer struct members by decreasing memory size

hebasto commented:
Probably @ mentions should be removed from the OP.

jonatack commented:
Yes. Removed @ mentions.

0xB10C commented:
ACK f639a88e720eb30c170e84710341b9c64ccf4ce4

Build, ran functional & unit tests, used `-netinfo` on testnet and signet.

practicalswift commented:
Concept ACK: I'm a big fan of `-netinfo` - thanks for improving!

0xB10C commented on src/bitcoin-cli.cpp:
```diff
@@ -491,11 +491,16 @@ class NetinfoRequestHandler : public BaseRequestHandler
-        // Report local addresses, ports, and scores.
-        result += "\nLocal addresses";
+        // Report reachable networks and local addresses info.
+        result += "\nnetworks:";
+        for (const UniValue& network : networkinfo["networks"].getValues()) {
+            const std::string name{network["name"].get_str()};
+            if (!name.empty() && network["reachable"].get_bool()) result += " " + name;
+        }
```
This produces the following output with `bitcoin-cli -testnet -netinfo` for me

```
Bitcoin Core v0.20.99.0-8b0aa43e2 testnet - 70016/Satoshi:0.20.99/

        ipv4    ipv6   onion   total  block-relay
in         0       0       0       0       0
out        9       0       1      10       2
total      9       0       1      10       2

networks: ipv4 ipv6 onion - local addresses:
2002:b10c:b10c:0:1234::5057     port  18333    score      1
2002:b10c:b10c:0:1234::33ed     port  18333    score      1
b10cabcdefaqqqhf.onion          port  18333    score     13
```

I'm not sure if this adds that much in its current form. Someone, who didn't look at the code, wouldn't know that `ipv4 ipv6 onion` is dynamically constructed from `networkinfo` and not just a constant. Additionally, it's for 1) quite hidden and 2) not self-explanatory that these are the reachable networks. 

Something like the following might be better:
```
reachable networks: ipv4 ipv6 onion
local addresses:
2002:b10c:b10c:0:1234::5057     port  18333    score      1
2002:b10c:b10c:0:1234::33ed     port  18333    score      1
b10cabcdefaqqqhf.onion          port  18333    score     13
```

jonatack replied:
Thanks for the testing and feedback, @0xB10C. I'm not sure it would add much either. Dropped the commit to not invalidate your previous ACK. Thanks!

jonatack committed:
cli -netinfo: make age column variable-width

as it has a wide possible range and the new name ("age" instead of "uptime") is
much shorter.

Emzy commented:
Tested ACK 43ee49f2c42f078763be55d666161507a2d7707b

Build and used -netinfo on mainnet, testnet and signet.

hebasto commented on src/bitcoin-cli.cpp:
```diff
@@ -419,13 +422,13 @@ class NetinfoRequestHandler : public BaseRequestHandler
         if (DetailsRequested() && !m_peers.empty()) {
             std::sort(m_peers.begin(), m_peers.end());
-            result += "Peer connections sorted by direction and min ping\n<-> relay   net  mping   ping send recv  txn  blk    age ";
+            result += strprintf("Peer connections sorted by direction and min ping\n<-> relay   net  mping   ping send recv  txn  blk %*s ", m_max_age_length, "age");
```
43ee49f2c42f078763be55d666161507a2d7707b

`#include <tinyformat.h>` ?

jonatack replied:
Well spotted. Will add it if I need to retouch or in the next PR to use the getpeerinfo connection types. There are quite a few `strprintf` in this file since years, so I guess it's not urgent.

jonatack replied:
included

jonatack replied:
Added the `tinyformat` header, and also `algorithm`, which was missing too. Thanks!

hebasto reviewed (approved):
ACK 43ee49f2c42f078763be55d666161507a2d7707b, tested on Linux Mint 20 (x86_64).

f3e30bf1b2231ba3abb65f59be1c5d923ca3dc91: isn't the https://github.com/bitcoin-core/bitcoin-devwiki/wiki/0.21.0-Release-Notes-Draft designed for this purpose?

~43ee49f2c42f078763be55d666161507a2d7707b: does `id` column require variable-width too?~

jonatack commented:
> f3e30bf1b2231ba3abb65f59be1c5d923ca3dc91: isn't the https://github.com/bitcoin-core/bitcoin-devwiki/wiki/0.21.0-Release-Notes-Draft designed for this purpose?

Yes, that wiki was not yet created when the PR was opened.

> 43ee49f2c42f078763be55d666161507a2d7707b: does `id` column require variable-width too?

Yes, it's already variable-width.

hebasto commented:
> > [43ee49f](https://github.com/bitcoin/bitcoin/commit/43ee49f2c42f078763be55d666161507a2d7707b): does `id` column require variable-width too?
> 
> Yes, it's already variable-width.

Indeed ^) Sorry for noise.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -398,10 +406,12 @@ class NetinfoRequestHandler : public BaseRequestHandler
+                m_peers.push_back({peer_id, mapped_as, version, last_blck, last_recv, last_send, last_trxn, min_ping, ping, addr, age, network, sub_version, is_block_relay, is_outbound});
                 m_max_id_length = std::max(ToString(peer_id).length(), m_max_id_length);
                 m_max_addr_length = std::max(addr.length() + 1, m_max_addr_length);
+                m_max_age_length = std::max(ToString(age).length(), m_max_age_length);
```
Hm, age is already a string.
```suggestion
                m_max_age_length = std::max(age.length(), m_max_age_length);
```

jonatack replied:
fixed

jonatack committed:
cli -netinfo: handle longer tor v3 local addresses

jonatack committed:
cli -netinfo: print oversized/extreme ping times as "-"

jonatack commented:
Thank you @Emzy, @hebasto, @0xB10C and @practicalswift for reviewing and testing. Updated per `git diff 43ee49f bab058e` to add the tinyformat header, remove an unneeded ToString type conversion, hoist the larger-sized Peer struct members to the start of the struct, and drop the release note.

michaelfolkson commented:
ACK 398045ba8b3694931069f88ec95553b3207dd1a6

Light code review, tested on MacOS Catalina with Signet and all 0-4 arguments.

A couple of (very) minor non-urgent observations not relevant for 0.21 (but maybe for a future follow up PR)

- I expected that as the arguments increased this would strictly increase the amount of the data displayed. Between 2 and 3 it replaces "address" with "version" only to display both address and version for 4. Was there particular thinking around this ie some users would want only one of these and not both?
- I suspect the help message for this tool could be sharpened up a touch.

Currently
```
 -netinfo
       Get network peer connection information from the remote server. An
       optional integer argument from 0 to 4 can be passed for **different
       peers listings** (default: 0).
```
Potential improvement
```
 -netinfo
       Get network peer connection information from the remote server. An
       optional integer argument from 0 to 4 can be passed **for a more
       detailed dashboard** (default: 0).
```

Emzy commented:
Tested ACK 398045ba8b3694931069f88ec95553b3207dd1a6

Build and used -netinfo on mainnet, testnet and signet on Ubuntu 20.04.1 LTS.

jonatack commented:
> Potential improvement
> 
> ```
>  -netinfo
>        Get network peer connection information from the remote server. An
>        optional integer argument from 0 to 4 can be passed **for a more
>        detailed dashboard** (default: 0).
> ```

Thanks @michaelfolkson, I'll do this in the next -netinfo update to use the `connection_type` field in the place of the first two columns.

jonatack commented:
Added the release note in the PR description to the wiki at https://github.com/bitcoin-core/bitcoin-devwiki/wiki/0.21.0-Release-Notes-Draft.
